File: src/main/scala/assignment3/e0/ResettableLatch.java
package assignment3.e0;

/*
 * Resettable Latch monitor.
 * 
 * - it is created with a number of participants
 * - like a latch, the gate is close until all participants signal
 * - the counter could be reset
 * 
 */
public class ResettableLatch {

	private int nParticipants;
	private int counter;
	
	public ResettableLatch(int nParticipants) {
		this.nParticipants = nParticipants;
		counter = nParticipants;
	}
	
	public synchronized void await() throws InterruptedException {
		while (counter > 0) {
			wait();
		} 
	}

	public synchronized void down() {
		counter--;
		if (counter == 0) {
			notifyAll();
		}
	}
	
	public synchronized void reset() {
		counter = nParticipants;
	}
	
}

File: src/main/scala/assignment3/e0/ResettableLatch.scala
package assignment3.e0

/*
 * Resettable Latch monitor.
 *
 * - it is created with a number of participants
 * - like a latch, the gate is close until all participants signal
 * - the counter could be reset
 *
 */
class ResettableLatch(var nParticipants: Int = 0) {
    private var counter = 0

    @throws[InterruptedException]
    def await(): Unit = synchronized{
      while (counter > 0) wait()
    }

    def down(): Unit = synchronized{
      counter -= 1
      if (counter == 0) notifyAll()
    }

    def reset(): Unit = synchronized{
      counter = nParticipants
    }
}